### PR TITLE
Investigations: unpivoting raw data and performance

### PR DESCRIPTION
**Pull request by ian-r-rose**
An investigation into #212 

I'm not happy with the performance of the upivoting operation (implemented in #124, solving #121). I *think* it should be a relatively cheap operation, since every "unpivot" is local to its row, and could be done on-the-fly in memory. But it seems that the Snowflake optimizer thinks otherwise! I took a bit of a deep dive into the performance characteristics of our five-minute-station-aggregate model, which has been the most expensive model in the project so far.

**Learning 1**: Unpivoting using the [`FLATTEN`](https://docs.snowflake.com/en/sql-reference/functions/flatten) function is shockingly slow. This is the approach taken in #124, because it allows you to more easily unpivot multiple columns at once.

**Learning 2**: Unpivoting using the [`UNPIVOT`](https://docs.snowflake.com/en/sql-reference/constructs/unpivot) function is similarly slow. Internally, it probably does something very similar to `flatten`. Snowflake does not document it, but duckdb [describes](https://duckdb.org/docs/sql/statements/unpivot.html#internals) something very similar, where unpivot is implemented as a series of `unnest` functions.

**Learning 3**: Despite my sense of wanting to unpivot as early in the model hierarchy as possible (so that _all_ downstream models see long data), it seems that it's worth it to unpivot *after* the five minute aggregation, simply because there is less data to sort at the end of the day.

This PR investigates two different unpivoting approaches, before and after aggregating the data, for a total of four combinations:

**Approach 1:** Unpivot using the `flatten` function, as is done in `main`.

**Approach 2:** *Manually unpivot by individually selecting the "wide" columns, naming them the same thing, and combining the subsets using `UNION ALL`.

The following are some timing data for the experiments in this PR. I ran everything on an XL warehouse, and numbers are in seconds:

|    | `FLATTEN` | `UNION ALL` |
|---|------------|--------------|
| Pre-agg |    33m 28s    |          3m 14s             |
| Post-agg | 15m 2s     |            91s       |

Clearly, it makes a huge difference! The approach of aggregating first, then using `UNION ALL` instead of `FLATTEN` cuts down the time by a factor of 20!

@tnrahim would you mind reviewing this to see if you agree with the analysis? If so, I think we should take this approach in our basal models. It will mean somewhat more complex/repetative code, but as you can see from this PR, we can cut down on the repetition using jinja.

### CODE REVIEW

**Comment by kengodleskidot** [comment]
I wanted to follow up to this topic by confirming if 14 fields/lanes are needed for flow, occupancy and speed at this time. Based on my review of the raw data tables in the data lake the various district schemas have 10 fields/lanes for flow, occupancy and speed.

File: transform/models/intermediate/db96/int_db96__five_minute_station_agg_flatten.sql
Comment (mmmiah) [inline transform/models/intermediate/db96/int_db96__five_minute_station_agg_flatten.sql:1]: is this lane number 14 is consistent over historical time period? If not, is there a way, we can determine n_lanes value based on availability of lane number like have ' _n' at the end of 'lane_'.

Comment (mmmiah) [inline transform/models/intermediate/db96/int_db96__five_minute_station_agg_flatten.sql:1]: Can we generate this list of number 'lateral flatten([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]) as lane' based on highest lane number?

File: transform/models/intermediate/db96/int_db96__five_minute_station_agg_unpivot.sql
Comment (mmmiah) [inline transform/models/intermediate/db96/int_db96__five_minute_station_agg_unpivot.sql:1]: same questions here: is this lane number 14 is consistent over historical time period. If not, is there a way, we can determine n_lanes value based on availability of lane number like have _n at the end of lane_n.

**Review by mmmiah — COMMENTED** [review]
Looks great! Left couple of comments!

**Comment by ian-r-rose** [comment]
Any updates here @tnrahim?

File: transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql [hunk 1]
```diff
@@ -2,10 +2,11 @@
     materialized="incremental",
     cluster_by=["sample_date"],
     unique_key=["id", "lane", "sample_timestamp"],
```
Comment (ian-r-rose) [inline transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql:4]: Let's also include `"sample_date"` among the unique keys to hint to the Snowflake optimizer that we can merge based on that criterion.

File: transform/models/intermediate/db96/int_db96__detector_agg_five_minutes.sql [hunk 2]
```diff
@@ -0,0 +1,91 @@
+{{ config(
+    materialized="incremental",
+    cluster_by="sample_date",
+    unique_key=["id", "lane", "sample_timestamp"],
```
Comment (ian-r-rose) [inline transform/models/intermediate/db96/int_db96__detector_agg_five_minutes.sql:4]: Same here, I think it may be helpful to the optimizer to include `"sample_date"` in the unique keys (even though it is redundant with the timestamp). I haven't done profiling on it, however, just a hunch that including the clustering key in merge criteria is helpful.

File: transform/models/intermediate/db96/int_db96__detector_agg_five_minutes.sql [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                and sample_date
+                >= dateadd(
+                    day,
+                    {{ var("dev_model_look_back") }},
+                    current_date()
+                )
+            {% endif %}
+    {% elif target.name != 'prd' %}
+        where sample_date >= dateadd(day, {{ var("dev_model_look_back") }}, current_date())
+    {% endif %}
```
Comment (ian-r-rose) [inline transform/models/intermediate/db96/int_db96__detector_agg_five_minutes.sql:46]: Let's use the `{{ make_model_incremental("sample_date") }}` macro here.

File: transform/models/intermediate/diagnostics/int_diagnostics__samples_per_station.sql [hunk 4]
```diff
@@ -9,7 +9,7 @@ with
 
 source as (
     select *
-    from {{ ref ('stg_clearinghouse__station_raw') }}
+    from {{ ref ('int_clearinghouse__detector_agg_five_minutes') }}
```
Comment (ian-r-rose) [inline transform/models/intermediate/diagnostics/int_diagnostics__samples_per_station.sql:12]: This is a major change to the meaning of this model, since this model involves many 30-second sample counts. This changes it to count 5-minute aggregate samples (effectively reducing by a factor of 10 the counts). Please revert this change and instead apply the unpivoting-with-jinja logic to this model.

File: transform/models/intermediate/performance/int_performance__detector_metrics_agg_five_minutes.sql [hunk 5]
```diff
@@ -7,8 +7,18 @@
...
+        sample_date,
+        sample_timestamp,
+        lane,
+        district,
+        avg(occupancy) as occupancy_avg,
+        sum(volume) as volume_sum,
+        sum(volume * speed) / nullifzero(sum(volume)) as speed_weighted
+    from {{ ref('int_clearinghouse__detector_agg_five_minutes') }}
     where {{ make_model_incremental('sample_date') }}
+    group by id, lane, sample_date, sample_timestamp, district
```
Comment (ian-r-rose) [inline transform/models/intermediate/performance/int_performance__detector_metrics_agg_five_minutes.sql:21]: This is another change to this downstream model, aggregating based on something that has already been aggregated. I'd like this PR to make changes to *only* the `int_clearinghouse__detector_agg_five_minutes` and `int_diagnostics__samples_per_station` models. To downstream models it should look 100% the same, we're just doing a performance optimization.

File: transform/models/intermediate/diagnostics/int_diagnostics__constant_occupancy.sql [hunk 6]
```diff
@@ -16,6 +16,18 @@ source as (
...
+    select
+        id,
+        sample_date,
+        sample_timestamp,
+        lane,
+        district,
+        avg(occupancy) as occupancy_avg
+    from source
+    group by id, lane, sample_date, sample_timestamp, district
+),
```
Comment (ian-r-rose) [inline transform/models/intermediate/diagnostics/int_diagnostics__constant_occupancy.sql:29]: This column is already averaged and grouped by these things (you did it in the `int_clearinghouse__detector_agg_five_minutes`). The issue here is that you changed the name of the occupancy column in the upstream model. For the most part, this PR should not need any changes to any downstream models based in the five-minute-aggregate model.

**Review by ian-r-rose — COMMENTED** [review]
Making good progress, but ideally we'd like to see *no* changes to downstream models: this is a performance optimization for `int_clearinghouse__detector_agg_five_minute` and `int_diagnostics__samples_per_station` where the resulting tables should look identical. Any changes to models downstream of that should be indicative of something that needs correcting.

**Comment by ian-r-rose** [comment]
@tnrahim I pushed a workaround here for the weird inconsistency in how sqlfluff was formatting `loop.last`. It seems that it's a bug others have run into, and is being tracked here: https://github.com/sqlfluff/sqlfluff/issues/5371

File: transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql [hunk 7]
```diff
@@ -20,27 +21,65 @@ with station_raw as (
...
+    count_if(volume is not null and occupancy is not null)
+        as sample_ct,
+        -- Sum of all the flow values
+    sum(volume) as volume_sum,
+    -- Average of all the occupancy values
+    avg(occupancy) as occupancy_avg,
+    -- calculate_weighted_speed
+    sum(volume * speed) / nullifzero(sum(volume)) as speed_weighted
+from agg_unioned
+group by id, lane, sample_date, sample_timestamp, district
```
Comment (ian-r-rose) [inline transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql:85]: This is a redundant aggregation, the jinja-SQL above already does exactly these (except for the `sample_ct` column, which looks like it's been dropped).

Making this table look the same as the previous one is mostly a matter of naming the columns the same thing, not of doing extra aggregations.

File: transform/models/intermediate/diagnostics/int_diagnostics__samples_per_station.sql
Comment (ian-r-rose) [inline transform/models/intermediate/diagnostics/int_diagnostics__samples_per_station.sql:1]: Nice! Looks like you're trying to get #213 in here as well?

**Review by ian-r-rose — COMMENTED** [review]
Other than including `"sample_date"` in the unique keys for the two five-minute-aggregates, this is looking quite close to me. Let's get CI passing and merge!

**Review by ian-r-rose — COMMENTED** [review]
@tnrahim I notice that there is some inconsistency in the use of "flow" vs "volume". They are basically synonymous, but more of our downstream models are using "volume". Can you rename the "flow" columns to "volume" columns for consistency?

It also looks like this has caused a problem or two in CI: https://california.getdbt.com/deploy/6/projects/10/runs/3783/

File: transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql (hunk 1 above)
Comment (tnrahim) [reply transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql]: Added sample date to unique keys

File: transform/models/intermediate/db96/int_db96__detector_agg_five_minutes.sql (hunk 2 above)
Comment (tnrahim) [reply transform/models/intermediate/db96/int_db96__detector_agg_five_minutes.sql]: Added sample date to unique keys

File: transform/models/intermediate/db96/int_db96__detector_agg_five_minutes.sql (hunk 3 above)
Comment (tnrahim) [reply transform/models/intermediate/db96/int_db96__detector_agg_five_minutes.sql]: Updated to use     where {{ make_model_incremental('sample_date') }}

File: transform/models/intermediate/diagnostics/int_diagnostics__constant_occupancy.sql (hunk 6 above)
Comment (tnrahim) [reply transform/models/intermediate/diagnostics/int_diagnostics__constant_occupancy.sql]: This change has been reversed. int_clearinghouse__detector_agg_five_minutes has been updated to flow into this model without issues.

File: transform/models/intermediate/diagnostics/int_diagnostics__samples_per_station.sql
Comment (tnrahim) [reply transform/models/intermediate/diagnostics/int_diagnostics__samples_per_station.sql:1]: yes! It made sense to as we have modified the stage to no longer have aggregated date.

File: transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql (hunk 7 above)
Comment (tnrahim) [reply transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql]: I've updated this logic to go in the correct spot. Apologize for committing this, I was just putting the names these while thinking things through.

File: transform/models/intermediate/db96/int_db96__detector_agg_five_minutes.sql [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+         buckets, it will be assigned to the one latter one due to
+         the floor() call.
+        */
+        dateadd(
+            'minute',
+            floor(minute(sample_timestamp) / 5) * 5,
+            trunc(sample_timestamp, 'hour')
+        ) as sample_timestamp_trunc
+    from {{ ref("stg_db96__vds30sec") }}
+    {% if is_incremental() %}
```
[the remark this answers is not in the document]
Comment (tnrahim) [reply transform/models/intermediate/db96/int_db96__detector_agg_five_minutes.sql]: Updated if is_incremental block in db96 agg_five_minutes with the macro make_model_incremental.sql in

File: transform/models/intermediate/diagnostics/int_diagnostics__samples_per_station.sql
[the remark this answers is not in the document]
Comment (tnrahim) [reply transform/models/intermediate/diagnostics/int_diagnostics__samples_per_station.sql:1]: I've updated samples_per_station to samples per detector and resolved downstream conflicts.

File: transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql [hunk 9]
```diff
@@ -20,27 +21,82 @@ with station_raw as (
...
+            {{ lane }} as lane,
+            volume_{{ lane }} as volume_sum,
+            zero_vol_ct_{{ lane }} as zero_vol_ct,
+            occupancy_{{ lane }} as occupancy_avg,
+            zero_occ_ct_{{ lane }} as zero_occ_ct,
+            zero_vol_pos_occ_ct_{{ lane }} as zero_vol_pos_occ_ct,
+            zero_occ_pos_vol_ct_{{ lane }} as zero_occ_pos_vol_ct,
+            high_volume_ct_{{ lane }} as high_volume_ct,
+            high_occupancy_ct_{{ lane }} as high_occupancy_ct,
+            speed_weighted_{{ lane }} as speed_weighted
```
Comment (ian-r-rose) [inline transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql:88]: I don't think these have been added to the YAML for the model.

**Review by ian-r-rose — COMMENTED** [review]
Code looks good to me, but I still see an issue in the CI run: https://california.getdbt.com/deploy/6/projects/10/runs/3847/

File: transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql (hunk 9 above)
Comment (tnrahim) [reply transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql:88]: I've added fields without descriptions to the yaml for now, I'm not sure why the CICD continues create this table incorrectly within TRANSFORM_DEV"."DBT_CLOUD_PR_15_230_CLEARINGHOUSE" though.

Comment (ian-r-rose) [reply transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql:89]: It's not being cloned incorrectly, this is expected behavior of incremental models (see the docs [here](https://docs.getdbt.com/docs/build/incremental-models#what-if-the-columns-of-my-incremental-model-change)). I've pushed a config tweak that should help with CI, though a full refresh will still be required for the new columns

Comment (ian-r-rose) [reply transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql:89]: CI is now passing

Comment (tnrahim) [reply transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql:89]: Thank you! Are there any other config changes needed to let DBT know a full refresh is necessary when merging? Or can we go ahead and merge the branch?

Comment (ian-r-rose) [reply transform/models/intermediate/clearinghouse/int_clearinghouse__detector_agg_five_minutes.sql:89]: We'll need to run a full refresh manually, I can handle that.

I think my last ask is to populate the `foobar` descriptions in the model yaml

**Review by ian-r-rose — COMMENTED** [review]
Looks good, thanks @tnrahim!

**Comment by britt-allen** [comment]
I'd love a share out on this issue @tnrahim @ian-r-rose this friday during mid sprint check in

**Comment by britt-allen** [comment]
or thursday rather, so the full caltrans team gets a show and tell of this completed issue too @tnrahim @ian-r-rose

**Comment by ian-r-rose** [comment]
What about during our sprint check-in this afternoon?

**Comment by britt-allen** [comment]
yes i just edited my comment lol, would love that
